velox/connectors/clp/search_lib/QueryRunner.cpp

```diff
@@ -0,0 +1,1286 @@
+#include "clp_s/search/SearchUtils.hpp"
+#include "clp_s/search/clp_search/EncodedVariableInterpreter.hpp"
+#include "clp_s/search/clp_search/Grep.hpp"
+
+#include "velox/vector/ComplexVector.h"
+
+#define eval(op, a, b) \
+  (((op) == FilterOperation::EQ) ? ((a) == (b)) : ((a) != (b)))
+
+using namespace clp_s;
+using namespace clp_s::search;
+using namespace clp_s::search::clp_search;
+using namespace simdjson;
+
+namespace facebook::velox::connector::clp::search_lib {
+void QueryRunner::init(
+    clp_s::SchemaReader* schema_reader,
+    int32_t schema_id,
+    std::unordered_map<int32_t, clp_s::BaseColumnReader*> const& column_map) {
+  m_schema_reader = schema_reader;
+  m_num_messages = schema_reader->get_num_messages();
+  m_clp_string_readers.clear();
+  m_var_string_readers.clear();
+  m_datestring_readers.clear();
+  m_basic_readers.clear();
+
+  auto matching_nodes_list = m_projection->get_matching_nodes_list();
+  for (const auto& node_ids : matching_nodes_list) {
+    if (node_ids.empty()) {
+      m_projected_columns.push_back(nullptr);
+      m_projected_types.push_back(NodeType::Unknown);
+      continue;
+    }
+
+    // Try to find a matching column in m_column_map
+    bool found_reader = false;
+    for (const auto node_id : node_ids) {
+      auto column_it = column_map.find(node_id);
+      if (column_it != column_map.end()) {
+        m_projected_columns.push_back(column_it->second);
+        m_projected_types.push_back(
+            m_schema_tree->get_node(node_id).get_type());
+        found_reader = true;
+        break;
+      }
+    }
+
+    if (!found_reader) {
+      m_projected_columns.push_back(nullptr);
+      m_projected_types.push_back(NodeType::Unknown);
+    }
+  }
+
+  for (auto& [column_id, column_reader] : column_map) {
+    if ((0 !=
+         (m_wildcard_type_mask &
+          node_to_literal_type(
+              m_schema_tree->get_node(column_id).get_type()))) ||
+        m_match->schema_searches_against_column(m_schema, column_id)) {
+      ClpStringColumnReader* clp_reader =
+          dynamic_cast<ClpStringColumnReader*>(column_reader);
+      VariableStringColumnReader* var_reader =
+          dynamic_cast<VariableStringColumnReader*>(column_reader);
+      DateStringColumnReader* date_reader =
+          dynamic_cast<DateStringColumnReader*>(column_reader);
+      if (nullptr != clp_reader &&
+          clp_reader->get_type() == NodeType::ClpString) {
+        m_clp_string_readers[column_id].push_back(clp_reader);
+      } else if (
+          nullptr != var_reader &&
+          var_reader->get_type() == NodeType::VarString) {
+        m_var_string_readers[column_id].push_back(var_reader);
+      } else if (nullptr != date_reader) {
+        // Datestring readers with a given column ID are guaranteed not to
+        // repeat
+        m_datestring_readers.emplace(column_id, date_reader);
+      } else {
+        m_basic_readers[column_id].push_back(column_reader);
+      }
```

Comment:
I saw you wrote like this somewhere else, so why not also do it here:
```suggestion
      if (auto* clp_reader = dynamic_cast<ClpStringColumnReader*>(column_reader);
          clp_reader && clp_reader->get_type() == NodeType::ClpString) {
        m_clp_string_readers[column_id].push_back(clp_reader);
      } else if (auto* var_reader = dynamic_cast<VariableStringColumnReader*>(column_reader);
                 var_reader && var_reader->get_type() == NodeType::VarString) {
        m_var_string_readers[column_id].push_back(var_reader);
      } else if (auto* date_reader = dynamic_cast<DateStringColumnReader*>(column_reader)) {
        m_datestring_readers.emplace(column_id, date_reader);
      } else {
        m_basic_readers[column_id].push_back(column_reader);
      }
```